Jaro-Winkler-Distance considers the length of strings, the number of matching characters and the number of transpositions, also we use a prefix scale that gives us more accurate answers when strings have a common prefix 

In [2]:
import logging

import jellyfish
import pandas as pd
import numpy as np

logger = logging.getLogger(__name__)

In [2]:
df_dbpedia = pd.read_csv("./files/dbpedia_classes.csv")
df_imageNet = pd.read_csv("./files/ImageNet.csv")
df_imageNet['Class Name'] = df_imageNet['Class Name'].str.replace(", ",",")
df_imageNet['Class Name'] = df_imageNet['Class Name'].str.split(",")

In [3]:
def get_jaro_winkler_distance():
    max_ratio_list = []
    max_label_list = []
    for class_list in df_imageNet['Class Name']:
        max_val = 0
        max_label = ''
        for item in class_list:
            for label in df_dbpedia['label']:
                if max_val< jellyfish.jaro_winkler_similarity(item, label):
                    max_val = jellyfish.jaro_winkler_similarity(item, label)
                    max_label = label
        max_ratio_list.append(max_val)
        max_label_list.append(max_label)
        
    df_imageNet['jaro_winkler_distance'] = max_ratio_list
    df_imageNet['dbpedia_class'] = max_label_list
    df_imageNet.rename(columns = {"Class Name":"imageNet_class"}, inplace=True)
            
get_jaro_winkler_distance()

In [4]:
df_imageNet


Class ID                                     imageNet_class  \
0           0                               [tench, Tinca tinca]   
1           1                      [goldfish, Carassius auratus]   
2           2  [great white shark, white shark, man-eater, ma...   
3           3                  [tiger shark, Galeocerdo cuvieri]   
4           4                     [hammerhead, hammerhead shark]   
..        ...                                                ...   
995       995                                        [earthstar]   
996       996  [hen-of-the-woods, hen of the woods, Polyporus...   
997       997                                           [bolete]   
998       998                            [ear, spike, capitulum]   
999       999     [toilet tissue, toilet paper, bathroom tissue]   

     jaro_winkler_distance  dbpedia_class  
0                 0.790000         tenure  
1                 0.693182    golf course  
2                 0.844444            man  
3                 0.762756  figure skater  
4                 0.777778         camera  
..                     ...            ...  
995               0.853333     earthquake  
996               0.715278         island  
997               0.800000           bone  
998               0.916667           year  
999               0.766026  tennis player  

[1000 rows x 4 columns]

Lets look at the number of classes with a distance score equal to 100

In [5]:
result =  df_imageNet[df_imageNet['jaro_winkler_distance' ]== 1.0]
result


Class ID                                     imageNet_class  \
134       134                                      [crane, bird]   
242       242                                            [boxer]   
286       286  [cougar, puma, catamount, mountain lion, paint...   
323       323  [monarch, monarch butterfly, milkweed butterfl...   
408       408                    [amphibian, amphibious vehicle]   
437       437         [beacon, lighthouse, beacon light, pharos]   
483       483                                           [castle]   
487       487  [cellular telephone, cellular phone, cellphone...   
497       497                          [church, church building]   
525       525                                  [dam, dike, dyke]   
553       553               [file, file cabinet, filing cabinet]   
624       624                                          [library]   
663       663                                        [monastery]   
668       668                                           [mosque]   
687       687                                [organ, pipe organ]   
705       705                   [passenger car, coach, carriage]   
743       743                             [prison, prison house]   
762       762   [restaurant, eating house, eating place, eatery]   
812       812                                    [space shuttle]   
829       829   [streetcar, tram, tramcar, trolley, trolley car]   
916       916           [web site, website, internet site, site]   
979       979                                     [valley, vale]   
980       980                                          [volcano]   
981       981                      [ballplayer, baseball player]   

     jaro_winkler_distance    dbpedia_class  
134                    1.0             bird  
242                    1.0            boxer  
286                    1.0          painter  
323                    1.0          monarch  
408                    1.0        amphibian  
437                    1.0       lighthouse  
483                    1.0           castle  
487                    1.0     mobile phone  
497                    1.0           church  
525                    1.0              dam  
553                    1.0             file  
624                    1.0          library  
663                    1.0        monastery  
668                    1.0           mosque  
687                    1.0            organ  
705                    1.0            coach  
743                    1.0           prison  
762                    1.0       restaurant  
812                    1.0    space shuttle  
829                    1.0        streetcar  
916                    1.0          website  
979                    1.0           valley  
980                    1.0          volcano  
981                    1.0  baseball player

In [6]:
print(result.shape)

(24, 4)


Lets look at the number of classes with a distance score higher than 0.9

In [7]:
result =  df_imageNet[df_imageNet['jaro_winkler_distance' ] > 0.9]
result

Class ID                                     imageNet_class  \
22         22  [bald eagle, American eagle, Haliaeetus leucoc...   
64         64                                      [green mamba]   
132       132  [American egret, great white heron, Egretta al...   
134       134                                      [crane, bird]   
231       231                                           [collie]   
..        ...                                                ...   
979       979                                     [valley, vale]   
980       980                                          [volcano]   
981       981                      [ballplayer, baseball player]   
989       989                           [hip, rose hip, rosehip]   
998       998                            [ear, spike, capitulum]   

     jaro_winkler_distance    dbpedia_class  
22                0.931429  American Leader  
64                0.905455       green alga  
132               0.914762  American Leader  
134               1.000000             bird  
231               0.909524          college  
..                     ...              ...  
979               1.000000           valley  
980               1.000000          volcano  
981               1.000000  baseball player  
989               0.916667             ship  
998               0.916667             year  

[103 rows x 4 columns]

Lets look at the number of classes with a distance score higher than 0.8

In [8]:
result =  df_imageNet[df_imageNet['jaro_winkler_distance' ] > 0.8]
result

Class ID                                     imageNet_class  \
2           2  [great white shark, white shark, man-eater, ma...   
5           5       [electric ray, crampfish, numbfish, torpedo]   
7           7                                             [cock]   
10         10              [brambling, Fringilla montifringilla]   
11         11                   [goldfinch, Carduelis carduelis]   
..        ...                                                ...   
988       988                                            [acorn]   
989       989                           [hip, rose hip, rosehip]   
990       990                  [buckeye, horse chestnut, conker]   
995       995                                        [earthstar]   
998       998                            [ear, spike, capitulum]   

     jaro_winkler_distance          dbpedia_class  
2                 0.844444                    man  
5                 0.841905  electrical substation  
7                 0.833333                   lock  
10                0.896296                  brain  
11                0.835133     Cardinal direction  
..                     ...                    ...  
988               0.893333                  actor  
989               0.916667                   ship  
990               0.894444                conifer  
995               0.853333             earthquake  
998               0.916667                   year  

[486 rows x 4 columns]

Lets look at the number of classes with a distance score higher than 0.7

In [9]:
result =  df_imageNet[df_imageNet['jaro_winkler_distance' ] > 0.7]
result

Class ID                                     imageNet_class  \
0           0                               [tench, Tinca tinca]   
2           2  [great white shark, white shark, man-eater, ma...   
3           3                  [tiger shark, Galeocerdo cuvieri]   
4           4                     [hammerhead, hammerhead shark]   
5           5       [electric ray, crampfish, numbfish, torpedo]   
..        ...                                                ...   
995       995                                        [earthstar]   
996       996  [hen-of-the-woods, hen of the woods, Polyporus...   
997       997                                           [bolete]   
998       998                            [ear, spike, capitulum]   
999       999     [toilet tissue, toilet paper, bathroom tissue]   

     jaro_winkler_distance          dbpedia_class  
0                 0.790000                 tenure  
2                 0.844444                    man  
3                 0.762756          figure skater  
4                 0.777778                 camera  
5                 0.841905  electrical substation  
..                     ...                    ...  
995               0.853333             earthquake  
996               0.715278                 island  
997               0.800000                   bone  
998               0.916667                   year  
999               0.766026          tennis player  

[910 rows x 4 columns]

Conclusion: We could use jaro_winkler distance but we need to set the treshold very high because with > 80 we already have 910 out of all classes.
A problem here is that it only compares strings not semantics (for example Class ID: 999 has a 0,76 distance but is actually not the same),
also there is no significant change between choosing jaro and jaro_winkler



Idea: Right now we take the highest value out of the image_net classes array, but lets change it so we take the average of the values as distance for each class and assign the class with the highest value to the dbpedia class

In [12]:
df_imageNet


Class ID                                     imageNet_class  \
0           0                               [tench, Tinca tinca]   
1           1                      [goldfish, Carassius auratus]   
2           2  [great white shark, white shark, man-eater, ma...   
3           3                  [tiger shark, Galeocerdo cuvieri]   
4           4                     [hammerhead, hammerhead shark]   
..        ...                                                ...   
995       995                                        [earthstar]   
996       996  [hen-of-the-woods, hen of the woods, Polyporus...   
997       997                                           [bolete]   
998       998                            [ear, spike, capitulum]   
999       999     [toilet tissue, toilet paper, bathroom tissue]   

     jaro_winkler_distance  dbpedia_class  
0                 0.790000         tenure  
1                 0.693182    golf course  
2                 0.844444            man  
3                 0.762756  figure skater  
4                 0.777778         camera  
..                     ...            ...  
995               0.853333     earthquake  
996               0.715278         island  
997               0.800000           bone  
998               0.916667           year  
999               0.766026  tennis player  

[1000 rows x 4 columns]

In [4]:
img_net = pd.read_csv("./files/ImageNet.csv")
img_net_list = img_net["Class Name"].tolist()

dbpedia = pd.read_csv("./files/dbpedia_classes.csv")
db_list = dbpedia["label"].tolist()

In [8]:
dataframe_jarow = pd.DataFrame(index =db_list ,columns =img_net_list)

In [9]:
for i in range(0,len(img_net_list)):
    for a in range(0,len(db_list)):
        if len(img_net_list[i].split(",")) > 1:
            max_val = 0.0
            for el in img_net_list[i].split(","):
                el_val = jellyfish.jaro_winkler_similarity(img_net_list[i], db_list[a])
                if el_val > max_val:
                    max_val = el_val
            dataframe_jarow.iat[a,i] = max_val
        else:
            dataframe_jarow.iat[a,i] = jellyfish.jaro_winkler_similarity(img_net_list[i], db_list[a])

In [10]:
dataframe_jarow

tench, Tinca tinca goldfish, Carassius auratus  \
company              0.420635                    0.453263   
activity             0.490741                    0.384259   
name                  0.37037                    0.429012   
person               0.481481                    0.425926   
actor                0.337037                    0.459259   
...                       ...                         ...   
Wind motor           0.457407                    0.495062   
Windmill             0.287037                    0.466049   
wine                 0.527778                    0.429012   
woman                0.337037                    0.491358   
zoo                       0.0                     0.45679   

           great white shark, white shark, man-eater, man-eating shark, Carcharodon caharias',  \
company                                              0.377128                                    
activity                                              0.49508                                    
name                                                 0.599398                                    
person                                               0.488286                                    
actor                                                 0.43427                                    
...                                                       ...                                    
Wind motor                                           0.446319                                    
Windmill                                             0.432731                                    
wine                                                 0.516064                                    
woman                                                0.532731                                    
zoo                                                       0.0                                    

           tiger shark, Galeocerdo cuvieri hammerhead, hammerhead shark  \
company                           0.391705                     0.285714   
activity                           0.37948                     0.386905   
name                              0.354839                     0.619048   
person                            0.598566                     0.468254   
actor                             0.476344                     0.490476   
...                                    ...                          ...   
Wind motor                        0.486738                     0.347619   
Windmill                          0.438172                      0.27381   
wine                              0.521505                     0.428571   
woman                             0.410753                      0.32381   
zoo                                    0.0                          0.0   

           electric ray, crampfish, numbfish, torpedo  stingray      cock  \
company                                      0.555556  0.490079  0.595238   
activity                                     0.464286  0.583333  0.458333   
name                                         0.496032  0.458333       0.0   
person                                       0.584127  0.513889       0.0   
actor                                        0.465079      0.55  0.483333   
...                                               ...       ...       ...   
Wind motor                                   0.406349  0.558333       0.0   
Windmill                                     0.365079       0.5       0.0   
wine                                         0.349206  0.583333       0.0   
woman                                        0.315873  0.383333  0.483333   
zoo                                               0.0       0.0  0.527778   

                 hen ostrich, Struthio camelus  ...  \
company          0.0                  0.288571  ...   
activity         0.0                     0.475  ...   
name             0.0                       0.0  ...   
person           0.5                  0.428889  ...   


In [12]:
a = dataframe_jarow
a = dataframe_jarow.replace("Not Present", np.nan)
a =a.replace(0.0, np.nan)

#Now we remove all columns where we have only NaN values
nan = a.dropna(axis=1, how="all")
highest_values_nan_max = nan.apply(lambda s: s.abs().nlargest(3).index.tolist(), axis=0)
highest_values_nan_max

tench, Tinca tinca goldfish, Carassius auratus  \
0  Concentration camp                        fish   
1      train carriage         Biological database   
2              engine      classical music artist   

  great white shark, white shark, man-eater, man-eating shark, Carcharodon caharias',  \
0                                            referee                                    
1                                             attack                                    
2                                          rest area                                    

  tiger shark, Galeocerdo cuvieri hammerhead, hammerhead shark  \
0               ice hockey league                       camera   
1            prehistorical period                         area   
2               historical period                      Caterer   

  electric ray, crampfish, numbfish, torpedo stingray               cock  \
0                                      lymph   stream               lock   
1                      electrical substation     star              coach   
2                        year in spaceflight   single  ice hockey player   

    hen ostrich, Struthio camelus  ... buckeye, horse chestnut, conker  \
0  chef      anatomical structure  ...                            beer   
1  vein           British royalty  ...                     hockey club   
2  fern         politician spouse  ...                          cheese   

  coral fungus    agaric   gyromitra stinkhorn, carrion fungus   earthstar  \
0        organ  aircraft      writer                     saint  earthquake   
1       colour   article  aristocrat     snooker world ranking      artist   
2     coal pit     award  motor race         Historical region  restaurant   

  hen-of-the-woods, hen of the woods, Polyporus frondosus, Grifola frondosa  \
0                                             cheese                          
1                                       chess player                          
2                                              hotel                          

    bolete ear, spike, capitulum  \
0     bone   year in spaceflight   
1   letter              ski area   
2  baronet             president   

        toilet tissue, toilet paper, bathroom tissue  
0  national collegiate athletic association team ...  
1                                             letter  
2   national collegiate athletic association athlete  

[3 rows x 1000 columns]

In [14]:
from pathlib import Path  

filepath = Path('../evaluation/files/jarow_winkler_top3.csv')  

filepath.parent.mkdir(parents=True, exist_ok=True)  

highest_values_nan_max.to_csv(filepath, index=False)  